### ISSUE TITLE
Circular dependency wrongly detect

### ISSUE THREAD

**Issue by tcassany** [hmlongco/Factory#58]
Hi,

With the resent version we have a circular dependency detected but I think it's wrong. It appears after the version `1.2.9`.

We have a container that exposes a `ParameterFactory` dependency. It provides different instances for different team. Each team has its own Container which provide an instance from the `ParameterFactory` dependency.
Depending on how we use Factory generate the instances, a circular dependency error is thrown.

Better than word to explain, here is a test to illustrate the issue

```
final class TeamLogger {
    let teamName: String

    init(teamName: String) {
        self.teamName = teamName
    }

    func log(_ message: String) {
        print("\(teamName) - \(message)")
    }
}

final class LoggerProvider {
    func logger(for team: String) -> TeamLogger {
        return TeamLogger(teamName: team)
    }
}

final class LoggerProviderContainer: SharedContainer {
    private static let loggerProvider = Factory(scope: .cached) {
        return LoggerProvider()
    }

    static let logger = ParameterFactory<String, TeamLogger> { teamName in
        return loggerProvider().logger(for: teamName)
    }
}

final class TeamAContainer: SharedContainer {
    static let teamALogger = Factory(scope: .cached) {
        LoggerProviderContainer.logger("TeamA")
    }
}


final class FactoryCircularDependencyTests: XCTestCase {

    override func setUp() {
        super.setUp()
        Container.Registrations.reset()
        Container.Scope.reset()
    }

    // ERROR 
    func testCircularDependencyError() {
        let logger = TeamAContainer.teamALogger()

        logger.log("Hello")
    }

    // OK
    func testNoCircularDependencyError() {
        let logger = LoggerProviderContainer.logger("TeamA")

        logger.log("Hello")
    }
}
```

`testCircularDependencyError` throws `Fatal error: circular dependency chain - FactoryTests.TeamLogger > FactoryTests.TeamLogger` .

**Comment by hmlongco** [comment]
Yeah, I can see that happening. There's a branch up called "chain" that may fix the problem if you want to try it out.

**Comment by tcassany** [comment]
Thanks. Chain branch solve the issue. 

Any idea when it will be merge to master?

**Comment by hmlongco** [comment]
Fixed in 1.3.7
